guides/md/writing_quantization_compatible_layers.md

```diff
@@ -0,0 +1,612 @@
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    def build(self, input_shape):
+        input_dim = input_shape[-1]
+        self._kernel = self.add_weight(
+            name="kernel",
+            shape=(input_dim,),
+            initializer="random_uniform",
+        )
+
+    def call(self, inputs, training=None):
+        return ops.multiply(inputs, self._kernel)
+
+    def quantize(self, mode, **kwargs):
+        if mode != "int8":
+            raise NotImplementedError(f"Unsupported quantization mode: {mode}")
+
+        quantized_kernel, scale = quantizers.abs_max_quantize(
+            self._kernel, axis=0, to_numpy=True
+        )
+        scale = ops.squeeze(scale, axis=0)
+
+        kernel_shape = self._kernel.shape
+
+        del self._kernel
+
+        self._int8_build(kernel_shape)
+
+        self._kernel.assign(quantized_kernel)
+        self.scale.assign(scale)
+
+        self._is_quantized = True
+
+        if self.dtype_policy.quantization_mode is None:
+            policy = dtype_policies.get(f"{mode}_from_{self.dtype_policy.name}")
+            self.dtype_policy = policy
+
+    def _int8_build(self, kernel_shape):
+        (input_dim,) = kernel_shape
+        self._kernel = self.add_weight(
+            name="kernel",
+            shape=(input_dim,),
+            initializer="zeros",
+            dtype="int8",
+            trainable=False,
+        )
+        self.scale = self.add_weight(
+            name="scale",
+            initializer="ones",
+            trainable=False,
+        )
+
+    def _int8_call(self, inputs, training=None):
+        x = ops.multiply(inputs, self._kernel)
+        x = ops.cast(x, self.compute_dtype)
+        x = ops.divide(x, self.scale)
+        return x
+
+```
+
+---
+## Try it: quantize and run a forward pass
+
+Below we build the layer, then quantize to INT8 and call it again.
+
+
+```python
+# Sample inputs
+rng = np.random.default_rng()
+x = rng.random((2, 4)).astype("float32")
+
+layer = SimpleScale()
+
+# Forward pass in float
+y_fp = layer(x)
+
+# Quantize to INT8 and run again
+layer.quantize("int8")
+y_int8 = layer(x)
+
+print("SimpleScale FP32 sample:", y_fp[0].numpy())
+print("SimpleScale INT8 sample:", y_int8[0].numpy())
+
+```
+
+<div class="k-default-codeblock">
+```
+SimpleScale FP32 sample: [-0.00359688  0.00296069 -0.00846314  0.00070467]
+SimpleScale INT8 sample: [-0.00359092  0.00290875 -0.00846319  0.00070462]
+```
+</div>
+
+---
+## Extending to INT4
+
+If you want to support INT4 quantization, add:
+
+- `_int4_build(...)`: allocate a packed 4-bit storage (often packed into int8) plus per-feature scales
+- `_int4_call(...)`: unpack at runtime and follow the same de-scale pattern
+- `quantize("int4")`: quantize weights with `value_range=(-8, 7)`, then pack to int4 storage
+
+See the
+[Dense](https://github.com/keras-team/keras/blob/3c3d6adc08db627d89b5ad5e7f9b0ba3e88f2641/keras/src/layers/core/dense.py)
```

Comment:
Because this link is for a specific version of `dense.py` you can safely put a line number anchor if you want to point to `_int4_call(...)` for instance.